### PR TITLE
Tests/test data class serializable dot dict

### PR DESCRIPTION
**Pull request by abegong**
Changes proposed in this pull request:
- Adds a set of test cases for typing objects using `@dataclass` and `SerializableDictDot` together.

@donaldheppner , you'd asked for good testing of the data types supporting `Expectation Diagnostic Reports`. As I was working, I found it more intuitive to do this in the abstract first, to understand how `@dataclass` and `SerializableDictDot` interact. With a handful of exceptions, the specific dataclasses required for `Expectation Diagnostics` are unremarkable implementation of the patterns set out here.

I've pulled this one commit out into its own PR, so that you can review it first. If this looks good to you, I'll use these patterns throughout the `Expectation Diagnostic Reports`. If there's something that you'd prefer to change, I can adjust accordingly.

Notes:
* This PR only covers type hinting. I'm not sure how best to implement type validation. Do we have a preferred pattern? My understanding is that we're using `marshmallow`, but our use is inconsistent. I would **love** to have a good reference pattern for this---I've already encountered several bugs that slip past type hinting, but would have been caught with type validation.
* I've also done some investigation of `SerializableDotDict` alongside `SerializableDictDot`. Their behavior is similar in most respects, but not identical. I suspect that they could be consolidated.
* I also have questions about the best way to handle serialization. This library looks promising, but I haven't investigated thoroughly: https://lidatong.github.io/dataclasses-json/. There may be other alternatives worth looking into.

### CODE REVIEW

**Comment by abegong** [comment]
Update:
* I've added a `SerializableDictDot.keys` method, to make its behavior more dictionary-like.
* Also some docstrings.

**Comment by abegong** [comment]
I've also added methods for
* `.to_dict`, with recursive type conversion.
* `.items`
* `.get`

With these tools, it starts to be pretty straightforward to substitute DotDict for dictionaries in the code, with pretty minimal changes to the code itself.

I've also experimented with using `pydantic` to do type validation. It's working really well. I'm pretty excited about it.

File: tests/expectations/test_dataclass_serializable_dot_dict_pattern.py [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+        ],
+        "enum_list": [
+            "x",
+            "y",
+            "z",
+        ]
+    }
+
+
+def test_immutability():
```
Comment (donaldheppner) [inline tests/expectations/test_dataclass_serializable_dot_dict_pattern.py]: I feel this test is verifying the python language (and @dataclass decorator), not our code

File: tests/expectations/test_dataclass_serializable_dot_dict_pattern.py [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+    my_F = MyClassF(
+        **{
+            "foo": "a string",
+            "bar": 1,
+            "in": 10,
+        }
+    )
+    assert my_F["foo"] == "a string"
+    assert my_F["bar"] == 1
+    assert my_F["input"] == 10
```
Comment (donaldheppner) [inline tests/expectations/test_dataclass_serializable_dot_dict_pattern.py]: add: assert my_F["in"] == 10

File: tests/expectations/test_dataclass_serializable_dot_dict_pattern.py [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+
+def test_basic_instantiation_with_arguments():
+    "Can be instantiated with arguments"
+    MyClassA(
+        foo="a string",
+        bar=1,
+    )
+
+
+def test_basic_instantiation_from_a_dictionary():
```
Comment (donaldheppner) [inline tests/expectations/test_dataclass_serializable_dot_dict_pattern.py]: Testing python, can be removed

File: tests/expectations/test_dataclass_serializable_dot_dict_pattern.py [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+        return len(self.A_list)
+
+    @property
+    def num_Bs(self):
+        return len(self.B_list)
+
+
+def test_basic_instantiation_with_arguments():
+    "Can be instantiated with arguments"
+    MyClassA(
```
Comment (donaldheppner) [inline tests/expectations/test_dataclass_serializable_dot_dict_pattern.py]: Testing python, can be removed

File: tests/expectations/test_dataclass_serializable_dot_dict_pattern.py [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+        **{
+            "foo": "a string",
+            "bar": 1,
+        }
+    )
+
+
+def test_access_uding_dot_notation():
+    "Keys can be accessed using dot notation"
+    my_A = MyClassA(
```
Comment (donaldheppner) [inline tests/expectations/test_dataclass_serializable_dot_dict_pattern.py]: Testing python, can be removed

File: great_expectations/types/__init__.py [hunk 6]
```diff
@@ -19,6 +22,44 @@ def __contains__(self, key):
...
+
+        # This is needed to play nice with pydantic.
+        if "__initialised__" in new_dict:
+            del new_dict["__initialised__"]
+
+        # DictDot's to_dict method works recursively, when a DictDot contains other DictDots.
+        for key, value in new_dict.items():
+            # Recursive conversion works on keys that are DictDots...
+            if isinstance(value, DictDot):
+                new_dict[key] = value.to_dict()
```
Comment (donaldheppner) [inline great_expectations/types/__init__.py:93]: do we need this method to be recursive and start with a deepcopy? Seems like we'd be creating a lot of copies of things

File: great_expectations/types/__init__.py [hunk 7]
```diff
@@ -19,6 +22,44 @@ def __contains__(self, key):
...
+
+            # ...and Enums...
+            elif isinstance(value, Enum):
+                new_dict[key] = value.value
+
+            # ...and when DictDots and Enums are nested one layer deeper in lists.
+            if isinstance(value, list):
```
Comment (donaldheppner) [inline great_expectations/types/__init__.py]: you may need to handle tuples as well

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: tests/expectations/test_dataclass_serializable_dot_dict_pattern.py (hunk 4 above)
Comment (abegong) [reply tests/expectations/test_dataclass_serializable_dot_dict_pattern.py]: Removing...

File: tests/expectations/test_dataclass_serializable_dot_dict_pattern.py (hunk 3 above)
Comment (abegong) [reply tests/expectations/test_dataclass_serializable_dot_dict_pattern.py]: Removing...

File: tests/expectations/test_dataclass_serializable_dot_dict_pattern.py (hunk 5 above)
Comment (abegong) [reply tests/expectations/test_dataclass_serializable_dot_dict_pattern.py]: Removing...

File: tests/expectations/test_dataclass_serializable_dot_dict_pattern.py (hunk 1 above)
Comment (abegong) [reply tests/expectations/test_dataclass_serializable_dot_dict_pattern.py]: Removing...

File: great_expectations/types/__init__.py (hunk 7 above)
Comment (abegong) [reply great_expectations/types/__init__.py]: Good point. I haven't actually encountered this yet...

File: tests/expectations/test_dataclass_serializable_dot_dict_pattern.py (hunk 2 above)
Comment (abegong) [reply tests/expectations/test_dataclass_serializable_dot_dict_pattern.py]: That won't actually work---and I don't believe that it should. See the updated test.

File: great_expectations/types/__init__.py (hunk 6 above)
Comment (abegong) [reply great_expectations/types/__init__.py:93]: I added a docstring to answer this question. Practically speaking, this makes a HUGE difference when trying to progressively type untyped parts of the codebase.

File: tests/expectations/test_dataclass_serializable_dot_dict_pattern.py (hunk 2 above)
Comment (donaldheppner) [reply tests/expectations/test_dataclass_serializable_dot_dict_pattern.py]: I'm worried we lose symmetry here: I gave you a dictionary with one set of keys, you're giving me back something different. How would the user know to access using ["input"] when they provided ["in"]?

File: great_expectations/types/__init__.py (hunk 7 above)
Comment (donaldheppner) [reply great_expectations/types/__init__.py]: I think this can be solved with a "or isinstance(value, tuple)"

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: great_expectations/types/__init__.py [hunk 8]
```diff
@@ -19,6 +65,49 @@ def __contains__(self, key):
...
```
Comment (abegong) [inline great_expectations/types/__init__.py]:
```suggestion
            if isinstance(value, list) or isinstance(value, tuple):
```

**Review by donaldheppner — APPROVED** [review]

**Pull request merged**